File: other_recursion/knapsack_modif.py






def knapsack(weights, max_weight):
    combs = []
    knapack_helper(weights, max_weight, 0, [], combs)
    return combs


def knapack_helper(weights, max_weight, idx, comb, combs):

    if max_weight < 0:
        return 
    if idx == len(weights):
        combs.append(comb.copy())
        return
    item_weight = weights[idx]
    comb.append(item_weight)
    knapack_helper(weights, max_weight-item_weight, idx+1, comb, combs)
    comb.pop()
    knapack_helper(weights, max_weight, idx+1, comb, combs)   


weights = [1, 2, 3]
max_weight = 3
res = knapsack(weights, max_weight)
print(f"For the following max_weight={max_weight} and weights={weights}")
print(f"possible combs are {res}")

File: other_recursion/knapsack_modif.js



function knapsack(weights, maxWeight) {
    let combs = [];
    knapsackHelper(weights, maxWeight, 0, [], combs);
    return combs;
}

function knapsackHelper(weights, maxWeight, idx, comb, combs){
    if (maxWeight < 0){
        return
    } 
    if (idx === weights.length){
        combs.push(comb.slice());
        return
    }
    let new_w = weights[idx];
    comb.push(new_w);
    knapsackHelper(weights, maxWeight-new_w, idx+1, comb, combs);
    comb.pop();
    knapsackHelper(weights, maxWeight, idx+1, comb, combs);
}



let weights = [1, 2, 3, 0.5];
let max_weight = 4;

let combs = knapsack(weights, max_weight);

console.log(`For the following max_weight=${max_weight} and weights=${JSON.stringify(weights)}`);
for (let comb of combs){
    console.log(`possible comb is ${comb}`);
}
